### PR TITLE
Decompile gf_gfx_loader.c

### PR DESCRIPTION
**Pull request by PikalaxALT**
TODO: Name gfx loading routines.

Disclosure: NNS types aped from leaked Nitro and NitroSys SDKs
Disclosure: Parameter type change in unk_02016B94.c informed in part by leaked pm_dp_ose source, needed to match typecast behavior in one function.

### CODE REVIEW

**Review by ProjectRevoTPP — DISMISSED** [review]

File: arm9/lib/include/GXS_ioreg.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
+#ifndef POKEDIAMOND_GXS_IOREG_H
+#define POKEDIAMOND_GXS_IOREG_H
+
+#define REG_GXS_DB_DISPCNT_O_SHIFT                         31
```
Comment (red031000) [inline arm9/lib/include/GXS_ioreg.h:4]: these should be in registers.h

File: arm9/lib/include/GX_ioreg.h [hunk 2]
```diff
@@ -0,0 +1,84 @@
+#ifndef NITRO_GX_IOREG_H
+#define NITRO_GX_IOREG_H
+
+#define REG_GX_DISPCNT_O_SHIFT                             31
```
Comment (red031000) [inline arm9/lib/include/GX_ioreg.h:4]: registers.h

File: arm9/lib/include/NNS_g2d.h [hunk 3]
```diff
@@ -0,0 +1,264 @@
+#ifndef NNSYS_NNS_G2D_H
+#define NNSYS_NNS_G2D_H
+#include "gx.h"
+
```
Comment (red031000) [inline arm9/lib/include/NNS_g2d.h:4]: honestly I don't really like the formatting in these files

File: arm9/lib/include/NNS_g2d.h [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+#define NNSYS_NNS_G2D_H
+#include "gx.h"
+
+typedef struct NNSG2dCharacterData
+{
+
+    u16                             H;
+    u16                             W;
+
+    GXTexFmt                        pixelFmt;    // GXTexFmt
```
Comment (red031000) [inline arm9/lib/include/NNS_g2d.h:11]: unneeded comments?

File: arm9/lib/include/NNS_g2d.h [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+    u32                                 szByteMax;             // Maximum byte count during all VRAM transfers
+    NNSG2dCellVramTransferData*         pCellTransferDataArray;// This is the length of the numCells amount of the maintained NNSG2dCellDataBank.
+    //
+
+}
+NNSG2dVramTransferData;
+
+typedef struct NNSG2dCellOAMAttrData
+{
+    u16         attr0;
```
Comment (red031000) [inline arm9/lib/include/NNS_g2d.h:166]: these are bitfields and probably should be treated as such (union with u16?)

File: arm9/src/gf_gfx_loader.c [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+THUMB_FUNC void * GfGfxLoader_UncompressFromNarc(NarcId narcId, s32 memberId, u32 heap_id)
+{
+    return GfGfxLoader_LoadFromNarc(narcId, memberId, TRUE, heap_id, FALSE);
+}
+
+THUMB_FUNC void * GfGfxLoader_LoadFromNarc(NarcId narcId, s32 memberNo, BOOL isCompressed, u32 heap_id, BOOL allocAtEnd)
+{
+    void * dest;
+    void * dest2;
+    if (isCompressed || allocAtEnd == TRUE)
```
Comment (red031000) [inline arm9/src/gf_gfx_loader.c:280]: == TRUE? is this needed for matching?

File: arm9/src/gf_gfx_loader.c [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+    {
+        u32 size = GetNarcMemberSizeByIdPair(narcId, memberNo);
+        dest = AllocFromHeap(heap_id, size);
+    }
+    if (dest != NULL)
+    {
+        ReadWholeNarcMemberByIdPair(dest, narcId, memberNo);
+        if (isCompressed)
+        {
+            if (allocAtEnd == FALSE)
```
Comment (red031000) [inline arm9/src/gf_gfx_loader.c:295]: again == FALSE

File: arm9/src/gf_gfx_loader.c [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+    {
+        u32 size = GetNarcMemberSizeByIdPair(narcId, memberNo);
+        dest = AllocFromHeapAtEnd(heap_id, size);
+    }
+    else
+    {
+        u32 size = GetNarcMemberSizeByIdPair(narcId, memberNo);
+        dest = AllocFromHeap(heap_id, size);
+    }
+    if (dest != NULL)
```
Comment (red031000) [inline arm9/src/gf_gfx_loader.c:290]: != NULL?

**Review by red031000 — CHANGES_REQUESTED** [review]

File: arm9/src/gf_gfx_loader.c (hunk 6 above)
Comment (PikalaxALT) [reply arm9/src/gf_gfx_loader.c:280]: Yes, without question. If the question related to `!= FALSE` it'd be a different question.

File: arm9/src/gf_gfx_loader.c (hunk 7 above)
Comment (PikalaxALT) [reply arm9/src/gf_gfx_loader.c:295]: Written this way for consistency with the above. May have been an enum in the original source.

File: arm9/src/gf_gfx_loader.c (hunk 8 above)
Comment (PikalaxALT) [reply arm9/src/gf_gfx_loader.c:290]: I prefer explicit comparison between pointers and NULL.

File: arm9/lib/include/NNS_g2d.h (hunk 5 above)
Comment (PikalaxALT) [reply arm9/lib/include/NNS_g2d.h:166]: Possibly. For now, I'm keeping SDK headers as-is.

**Review by red031000 — APPROVED** [review]

**Pull request merged**